### PR TITLE
Compare tagged/niche-filling layout and pick the best one

### LINKED ISSUE #63866: Niche-filling layouts are picked over tagged ones even when detrimental.
This prints the values in comments:

```rust
use std::num::NonZeroU8;

pub enum Foo { A(NonZeroU8, u32), B }
pub enum Bar { A(u8, u32), B }

fn main() {
    use std::mem::size_of;
    
    dbg!(size_of::<Foo>()); // = 8
    dbg!(size_of::<Option<Foo>>()); // = 12
    dbg!(size_of::<Option<Option<Foo>>>()); // = 12
    
    dbg!(size_of::<Bar>()); // = 8
    dbg!(size_of::<Option<Bar>>()); // = 8
    dbg!(size_of::<Option<Option<Bar>>>()); // = 8
}
```

Note that `Foo` contains *strictly more* opportunities for optimization, but it's optimized worse.

That's because it uses a "niche-filling" layout (using `0` in the `NonZeroU8` field of `A` for `B`), but the "tag" layout wouldn't be any larger, because it has enough space in `8` bytes to fit the (byte-sized) tag, `NonZeroU8` and `u32`.

`Bar` uses the tagged layout since it has no other choice, and ends up with `254` free values in the byte-sized tag, but `Foo` "jumps the gun" on using a pre-existing niche.

### PR DESCRIPTION
**Pull request by erikdesjardins**
Finishes up #71045, and so fixes #63866.

cc @eddyb
r? @nikomatsakis (since @eddyb wrote the first commit)

### CODE REVIEW

File: src/test/ui/type-sizes.rs [hunk 1]
```diff
@@ -102,6 +103,15 @@ enum Option2<A, B> {
     None
 }
 
+pub enum CanBeNicheFilledButShouldnt {
```
Comment (nikomatsakis) [inline src/test/ui/type-sizes.rs:106]: Can you add a comment explaining why it should not use a niche?

File: src/test/ui/type-sizes.rs [hunk 2]
```diff
@@ -102,6 +103,15 @@ enum Option2<A, B> {
...
+    A(NonZeroU8, u32),
+    B
+}
+pub enum AlwaysTagged {
```
Comment (nikomatsakis) [inline src/test/ui/type-sizes.rs:110]:
```suggestion
pub enum AlwaysTaggedBecauseItHasNoNiche {
```

maybe make this self-documenting?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
r=me after nits are fixed

**Review by nikomatsakis — APPROVED** [review]

**Review by eddyb — APPROVED** [review]

**Comment by eddyb** [comment]
Thanks a lot! I've closed my original unfinished PR.

**Comment by Mark-Simulacrum** [comment]
@rust-timer make-pr-for e775b4dbc4a991d60014225088c58eddd91f18cb

**Comment by Mark-Simulacrum** [comment]
Benchmarked in https://github.com/rust-lang/rust/pull/74592, it does appear like this was a 5-10% performance regression on a large variety of crates. We should probably revert this PR. I've posted a revert here: xxx

https://perf.rust-lang.org/compare.html?start=d3df8512d2c2afc6d2e7d8b5b951dd7f2ad77b02&end=cfade73820883adf654fe34fd6b0b03a99458a51

cc @nnethercote

**Comment by Mark-Simulacrum** [comment]
This turned out to not be the source of any regressions, and indeed might be a mild improvement. It will be re-landed in https://github.com/rust-lang/rust/pull/74802. See also https://github.com/rust-lang/rust/pull/74716#issuecomment-663735771 and the next few comments for discussion on this topic.

**Pull request merged**